File: src/No648.js
class Tire{

    constructor() {
        this.flag = false;
        this.word = null;
        this.children = {}
    }
}


let root = new Tire();

let dictionary = ["e","k","c","harqp","h","gsafc","vn","lqp","soy","mr","x","iitgm","sb","oo","spj","gwmly","iu","z","f","ha","vds","v","vpx","fir","t","xo","apifm","tlznm","kkv","nxyud","j","qp","omn","zoxp","mutu","i","nxth","dwuer","sadl","pv","w","mding","mubem","xsmwc","vl","farov","twfmq","ljhmr","q","bbzs","kd","kwc","a","buq","sm","yi","nypa","xwz","si","amqx","iy","eb","qvgt","twy","rf","dc","utt","mxjfu","hm","trz","lzh","lref","qbx","fmemr","gil","go","qggh","uud","trnhf","gels","dfdq","qzkx","qxw"]
let sentence = "s"

let lenDictionary = dictionary.length;

for(let i = 0; i < lenDictionary; i ++){
    let word = dictionary[i];
    let lenWord = dictionary[i].length;
    let cur = root;
    for(let j = 0; j < lenWord; j ++){
        if(cur.children[word[j].charCodeAt() - 'a'.charCodeAt()] == null){
            cur.children[word[j].charCodeAt() - 'a'.charCodeAt()] = new Tire();
        }
        cur = cur.children[word[j].charCodeAt() - 'a'.charCodeAt()];
    }
    cur.flag = true;
    cur.word = word;
}

let words = sentence.split(" ");
let ans = []
for(let word of words){
    let cur = root;
    let lenWord = word.length;
    let f = false;
    for(let i = 0; i < lenWord; i ++){
        if(cur.children[word[i].charCodeAt() - 'a'.charCodeAt()] != null){
            cur = cur.children[word[i].charCodeAt() - 'a'.charCodeAt()]
        }else{
            break;
        }
        if(cur.flag == true){
            f = true;
            ans.push(cur.word);
            break;
        }
    }
    if(!f){
        ans.push(word)
    }
}
console.log(ans.join(" "))

File: src/No648.java
class Tire{
    boolean flag;
    String word;
    Tire[] children;

    public Tire(){
        this.flag = false;
        this.word = null;
        this.children = new Tire[26];
    }

}

public class No648 {
    public static void main(String[] args) {
        String[] dictionary = {"a", "aa", "aaa", "aaaa"};
        int len = dictionary.length;
        Tire root = new Tire();
        for(int i = 0; i < len; i ++){
            String word = dictionary[i];
            int wordLen = word.length();
            Tire cur = root;
            for(int j = 0; j < wordLen; j ++){
                if(cur.children[word.charAt(j)-'a'] == null){
                    cur.children[word.charAt(j)-'a'] = new Tire();
                }
                cur = cur.children[word.charAt(j)-'a'];
            }
            cur.flag = true;
            cur.word = word;
        }
        String sentence = "a aa a aaaa aaa aaa aaa aaaaaa bbb baba ababa";
        StringBuilder sb = new StringBuilder();
        String[] words = sentence.split(" ");
        for(String word : words){
            int wordLen = word.length();
            Tire cur = root;
            boolean flag = false;
            for(int i = 0; i < wordLen; i ++){
                if(cur.children[word.charAt(i)-'a'] != null){
                    cur = cur.children[word.charAt(i)-'a'];
                }else {
                    break;
                }
                if(cur.flag == true){
                    sb.append(cur.word);
                    sb.append(" ");
                    flag = true;
                    break;
                }
            }
            if(!flag){
                sb.append(word);
                sb.append(" ");
            }
        }
        sb.deleteCharAt(sb.length()-1);
        System.out.println(sb.toString());
    }
}
